cupy/cuda/device.pyx

```diff
@@ -25,6 +24,43 @@ cpdef int get_device_id() except? -1:
     return runtime.getDevice()
 
 
+cdef class _ThreadLocalStack:
+
+    cdef object _local
+
+    def __init__(self):
+        self._local = threading.local()
```

Comment:
This is not correct unfortunately. What's done here will create the `_ThreadLocal` instance only in the thread that initializes this module; other new threads do not have this object. 

Please follow the practice done elsewhere and define a factory method `get()` so that each thread has its own copy of `_ThreadLocal`:
https://github.com/cupy/cupy/blob/596d1af53b5793d3d52994c8b493ff42be453a8d/cupy_backends/cuda/stream.pyx#L16-L22

Reply:
It is correct as written. The thread-local object itself is visible to all threads. Its attributes vary depending on which thread accesses them. The equivalent of the idiom you're pointing to here is handled further down where I do
```
current = getattr(self._local, "_device_stack", None)
if current is None:
    current = []
    self._local._device_stack = current
```
In your try-except example the try is to detect whether or not the attribute on the thread-local object is present. The thread-local object is present in all threads. That's why it's catching an AttributeError, not a NameError.

Reply:
I guess I can try to refactor to more closely mimic the idiom used in the stream file if that'd help.

Reply:
No please don't. I really had a bad day and wasn't thinking straight. I think you are correct. Sorry for the disturbance.

Reply:
Okay, if I _exactly_ follow the idiom in the other file we end up implicitly creating an unnecessary Cython vtable implicitly. I've pushed a version that does that though.

Reply:
Thanks! Sorry again for my unreasonable requests.

Reply:
No problem. Thanks for taking the time to review the various versions!